### PR TITLE
move registries to provider interface

### PR DESCRIPTION
**Pull request by ddollar**
Moves registries into the provider interface.

* `docker login` is now only ever called during `convox build`
* App servers no longer try to login to private registries on boot
* Solves a token expiration issue that was showing up after API instances had been up for a few hours

### CODE REVIEW

File: provider/mock_Provider.go [hunk 1]
```diff
@@ -0,0 +1,1130 @@
+package provider
+
+import io "io"
+import mock "github.com/stretchr/testify/mock"
+import structs "github.com/convox/rack/api/structs"
+
+// MockProvider is an autogenerated mock type for the Provider type
+type MockProvider struct {
```
Comment (MiguelMoll) [inline provider/mock_Provider.go]: +1 for MockProvider

File: client/registries.go [hunk 2]
```diff
@@ -23,34 +18,41 @@ func (c *Client) AddRegistry(server, username, password, email string) (*Registr
...
+	system, err := c.GetSystem()
 	if err != nil {
 		return nil, err
 	}
 
-	return &registry, nil
-}
+	// backwards compatible
+	// FIXME: pick the right version here
+	if system.Version < "30000000000000" {
```
Comment (MiguelMoll) [inline client/registries.go]: Acceptable to go out like this?

File: api/main.go [hunk 3]
```diff
@@ -17,12 +13,5 @@ func recoverWith(f func(err error)) {
 }
 
 func main() {
-	// prime the API and instance for builds by logging into private registries
-	// and pulling down latest app images
-	go func() {
-		models.LoginPrivateRegistries()
-		models.PullAppImages()
```
Comment (MiguelMoll) [inline api/main.go]: Build cache will be handled elsewhere?

File: client/registries.go (hunk 2 above)
Comment (ddollar) [inline client/registries.go]: Yes, I'll fix it in a quick CLI release once I have a release number for this release.

File: api/main.go (hunk 3 above)
Comment (ddollar) [inline api/main.go]: It's handled by Docker. Pulling the images on the API instance doesn't necessarily help anymore anyway.

**Pull request closed without merging**